File: framework/ndn_iot_js/js/discovery/sync-based-discovery.js
// Generic sync-based discovery implementation
var SyncBasedDiscovery = function SyncBasedDiscovery
  (face, keyChain, certificateName, syncPrefix, observer, serializer, 
   syncDataFreshnessPeriod, initialDigest, syncInterestLifetime, syncInterestMinInterval,
   timeoutCntThreshold, maxResponseWaitPeriod, minResponseWaitPeriod, entityDataFreshnessPeriod)
{
    if (syncDataFreshnessPeriod === undefined) {
        syncDataFreshnessPeriod = 4000;
    }
    if (initialDigest === undefined) {
        initialDigest = "00";
    }
    if (syncInterestLifetime === undefined) {
        syncInterestLifetime = 4000;
    }
    if (syncInterestMinInterval === undefined) {
        syncInterestMinInterval = 500;
    }
    if (timeoutCntThreshold === undefined) {
        timeoutCntThreshold = 3;
    }
    if (maxResponseWaitPeriod === undefined) {
        maxResponseWaitPeriod = 2000;
    }
    if (minResponseWaitPeriod === undefined) {
        minResponseWaitPeriod = 400;
    }
    if (entityDataFreshnessPeriod === undefined) {
        entityDataFreshnessPeriod = 10000;
    }

    this.face = face;
    this.keyChain = keyChain;
    this.syncPrefix = syncPrefix;
    
    this.objects = {};
    this.hostedObjects = {};

    this.memoryContentCache = new MemoryContentCache(this.face);
    this.certificateName = new Name(certificateName);

    this.currentDigest = initialDigest;
    this.syncDataFreshnessPeriod = syncDataFreshnessPeriod;
    this.initialDigest = initialDigest;
    this.syncInterestLifetime = syncInterestLifetime;

    this.syncInterestMinInterval = syncInterestMinInterval;
    this.timeoutCntThreshold = timeoutCntThreshold;
    this.entityDataFreshnessPeriod = entityDataFreshnessPeriod;

    this.observer = observer;
    this.serializer = serializer;

    return;
}

//Public facing interface
SyncBasedDiscovery.prototype.start = function ()
{
    this.updateDigest();
    var interest = new Interest((new Name(this.syncPrefix)).append(this.currentDigest));

    interest.setMustBeFresh(true);
    interest.setInterestLifetimeMilliseconds(this.syncInterestLifetime);
    this.face.expressInterest(interest, this.onSyncData.bind(this), this.onSyncTimeout.bind(this));

    console.log("Express interest: " + interest.getName().toUri());
    return;
}

SyncBasedDiscovery.prototype.stop = function ()
{
    this.memoryContentCache.unregisterAll();
    return;
}


SyncBasedDiscovery.prototype.getHostedObjects = function ()
{
    return this.hostedObjects;
}

SyncBasedDiscovery.prototype.getObjects = function ()
{
    return this.objects;
}

SyncBasedDiscovery.prototype.addHostedObject = function
  (name, entityInfo)
{
    // If this is the first object we host, we register for sync namespace: meaning a participant not hosting anything 
    //   is only "listening" for sync, and will not help in the sync process
    if (Object.keys(this.hostedObjects).length == 0) {
        this.memoryContentCache.registerPrefix(this.syncPrefix, this.onRegisterFailed.bind(this), this.onSyncInterest.bind(this));
    }
    if (this.addObject(name)) {
        this.hostedObjects[name] = entityInfo;
        // TODO: debug this, seems to not working as intended yet
        this.contentCacheAddEntityData(name, entityInfo);
        this.memoryContentCache.registerPrefix(new Name(name), this.onRegisterFailed.bind(this), this.onEntityDataNotFound.bind(this));
    } else {
        console.log("Item with this name already added");
    }

    return;
}

SyncBasedDiscovery.prototype.removeHostedObject = function (name)
{
    if (name in this.hostedObjects) {
        delete this.hostedObjects[name];
        if (Object.keys(this.hostedObjects).length == 0) {
            self._memoryContentCache.unregisterAll();
        }
        if (this.removeObject(name)) {
            return true;
        } else {
            console.log("Hosted item not in objects list");
            return false;
        }
    } else {
        return false;
    }
}

//Internal functions
SyncBasedDiscovery.prototype.contentCacheAddEntityData = function
  (name, entityInfo)
{
    var content = this.serializer.serialize(entityInfo);
    var data = new Data(new Name(name));

    data.setContent(content);
    // Interest issuer should not ask for mustBeFresh in this case, for now
    data.getMetaInfo().setFreshnessPeriod(this.entityDataFreshnessPeriod);

    var self = this;
    this.keyChain.sign(data, this.certificateName, function() {
        self.memoryContentCache.add(data);
    });
}

SyncBasedDiscovery.prototype.contentCacheAddSyncData = function
  (dataName)
{
    var keys = Object.keys(this.objects);
    keys.sort();

    var content = "";
    for (var item in keys) {
        content += keys[item] + "\n";
    }
    content.trim();

    var data = new Data(new Name(dataName));
        
    data.setContent(content);
    data.getMetaInfo().setFreshnessPeriod(this.syncDataFreshnessPeriod);
    
    var self = this;
    this.keyChain.sign(data, this.certificateNamem, function() {
        // adding this data to memoryContentCache should satisfy the pending interest
        self.memoryContentCache.add(data);
    });
}
        

SyncBasedDiscovery.prototype.onSyncInterest = function
  (prefix, interest, face, interestFilterId, filter)
{
    if (interest.getName().size() !== this.syncPrefix.size() + 1) {
        // Not an interest for us
        return;
    }
            
    var digest = interest.getName().get(-1).toEscapedString();
    this.updateDigest();
    if (digest != this.currentDigest) {
        // Wait a random period before replying; rationale being that "we are always doing ChronoSync recovery...this is the recovery timer but randomized"
        // Consider this statement: we are always doing ChronoSync recovery
        // TODO: this has the problem of potentially answering with wrong data, there will be more interest exchanges needed for the lifetime duration of one wrong answer
        // Consider appending "answerer" as the last component of data name?
        // TODO2: don't see why we should wait here

        this.replySyncInterest(interest, digest);
        //dummyInterest = Interest(Name("/local/timeout1"))
        //dummyInterest.setInterestLifetimeMilliseconds(random.randint(self._minResponseWaitPeriod, self._maxResponseWaitPeriod))
        //self._face.expressInterest(dummyInterest, self.onDummyData, lambda a : self.replySyncInterest(a, digest))
    }

    return;
}
        

SyncBasedDiscovery.prototype.replySyncInterest = function
  (interest, receivedDigest)
{
    this.updateDigest();
    if (receivedDigest != self._currentDigest) {
        // TODO: one participant may be answering with wrong info: scenario: 1 has {a}, 2 has {b}
        // 2 gets 1's {a} and asks again before 1 gets 2's {b}, 2 asks 1 with the digest of {a, b}, 1 will 
        // create a data with the content {a} for the digest of {a, b}, and this data will be able to answer
        // later steady state interests from 2 until it expires (and by which time 1 should be updated with
        // {a, b} as well)
        this.contentCacheAddSyncData(new Name(this.syncPrefix).append(receivedDigest));
    }
    return;
}

SyncBasedDiscovery.prototype.onSyncData = function
  (interest, data)
{
    //TODO: do verification first
    console.log("Got sync data; name: " + data.getName().toUri() + "; content: " + data.getContent().buf());
    var content = String(data.getContent().buf()).split('\n');

    for (var itemName in content) {
        if (!(content[itemName] in this.objects)) {
            if (content[itemName] != "") {
                this.onReceivedSyncData(content[itemName]);
            }
        }
    }
    
    // Hack for re-expressing sync interest after a short interval
    var dummyInterest = new Interest(new Name("/local/timeout"));
    dummyInterest.setInterestLifetimeMilliseconds(this.syncInterestLifetime);
    this.face.expressInterest(dummyInterest, this.onDummyData.bind(this), this.expressSyncInterest.bind(this));
    return;
}

SyncBasedDiscovery.prototype.onSyncTimeout = function
  (interest)
{
    console.log("Sync interest times out: " + interest.getName().toUri());
    var newInterest = new Interest(new Name(this.syncPrefix).append(this.currentDigest));
    newInterest.setInterestLifetimeMilliseconds(this.syncInterestLifetime);
    newInterest.setMustBeFresh(true);
    this.face.expressInterest(newInterest, this.onSyncData.bind(this), this.onSyncTimeout.bind(this));
    console.log("Express interest: " + newInterest.getName().toUri());

    return;
}

// Handling received sync data: express entity interest
SyncBasedDiscovery.prototype.onReceivedSyncData = function 
  (itemName)
{
    console.log("Received itemName: " + itemName);
    var interest = new Interest(new Name(itemName));
    interest.setInterestLifetimeMilliseconds(4000);
    interest.setMustBeFresh(false);
    this.face.expressInterest(interest, this.onEntityData.bind(this), this.onEntityTimeout.bind(this));
    
    return;
}

SyncBasedDiscovery.prototype.onEntityTimeout = function 
  (interest)
{
    console.log("Item interest times out: " + interest.getName().toUri());
    return;
}

SyncBasedDiscovery.prototype.onEntityData = function 
  (interest, data)
{
    var self = this;
    console.log("Got data: " + data.getName().toUri());
    this.addObject(interest.getName().toUri());
    console.log("Added device: " + interest.getName().toUri());

    var dummyInterest = new Interest(new Name("/local/timeout"));
    dummyInterest.setInterestLifetimeMilliseconds(4000);
    this.face.expressInterest(dummyInterest, this.onDummyData.bind(this), function (a) {
        self.expressHeartbeatInterest(a, interest)
    });
    return;
}

SyncBasedDiscovery.prototype.expressHeartbeatInterest = function
  (dummyInterest, entityInterest)
{
    var newInterest = new Interest(entityInterest);
    newInterest.refreshNonce();

    this.face.expressInterest(entityInterest, this.onHeartbeatData.bind(this), this.onHeartbeatTimeout.bind(this)); 
}

SyncBasedDiscovery.prototype.onHeartbeatData = function
  (interest, data)
{
    var self = this;
    this.resetTimeoutCnt(interest.getName().toUri());
    var dummyInterest = new Interest(new Name("/local/timeout"));
    dummyInterest.setInterestLifetimeMilliseconds(4000);
    this.face.expressInterest(dummyInterest, this.onDummyData.bind(this), function (a) {
        self.expressHeartbeatInterest(a, interest);
    });
}

SyncBasedDiscovery.prototype.onHeartbeatTimeout = function 
  (interest)
{
    if (this.incrementTimeoutCnt(interest.getName().toUri())) {
        console.log("Remove: " + interest.getName().toUri() + " because of consecutive timeout cnt exceeded");
    } else {
        var newInterest = new Interest(interest.getName());
        console.log("Express interest: " + newInterest.getName().toUri());
        newInterest.setInterestLifetimeMilliseconds(4000);
        this.face.expressInterest(newInterest, this.onHeartbeatData.bind(this), this.onHeartbeatTimeout.bind(this));
    }
}

SyncBasedDiscovery.prototype.onDummyData = function
  (interest, data)
{
    console.log("Unexpected reply to dummy interest: " + data.getContent().buf());
    return;
}

SyncBasedDiscovery.prototype.expressSyncInterest = function
  (interest)
{
    var newInterest = new Interest(new Name(this.syncPrefix).append(this.currentDigest));
    newInterest.setInterestLifetimeMilliseconds(this.syncInterestLifetime);
    newInterest.setMustBeFresh(true);
    this.face.expressInterest(newInterest, this.onSyncData.bind(this), this.onSyncTimeout.bind(this));
    console.log("Dummy timeout; Express interest: " + newInterest.getName().toUri());
    return;
}

SyncBasedDiscovery.prototype.addObject = function 
  (name)
{
    if (name in this.objects) {
        return false;
    } else {
        this.objects[name] = {"timeout_count": 0};
        this.notifyObserver(name, "ADD", "");
        this.contentCacheAddSyncData(new Name(this.syncPrefix).append(this.currentDigest));
        this.updateDigest();
        return true;
    }
}

SyncBasedDiscovery.prototype.removeObject = function
  (name)
{
    if (name in this.objects) {
        delete self._objects[name]
        
        this.notifyObserver(name, "REMOVE", "");
        this.contentCacheAddSyncData(new Name(this.syncPrefix).append(this.currentDigest));
        this.updateDigest();
        return true;
    } else {
        return false;
    }
}

SyncBasedDiscovery.prototype.updateDigest = function ()
{
    // TODO: for now, may change the format of the list encoding for easier cross language compatibility
    var keys = Object.keys(this.objects);
    keys.sort();

    if (keys.length > 0) {
        var m = Crypto.createHash('sha256');
        for (var i = 0; i < keys.length; i++) {
            m.md.updateString(keys[i]);
        }
        this.currentDigest = m.digest('hex');
    } else {
        this.currentDigest = this.initialDigest;
    }
    return;
}
        
SyncBasedDiscovery.prototype.incrementTimeoutCnt = function 
  (name)
{
    if (name in this.objects) {
        this.objects[name]["timeout_count"] += 1;
        if (this.objects[name]["timeout_count"] >= this.timeoutCntThreshold) {
            return this.removeObject(name);
        } else {
            return false;
        }
    } else {
        return false;
    }
}  

SyncBasedDiscovery.prototype.resetTimeoutCnt = function
  (name)
{
    if (name in this.objects) {
        this.objects[name]["timeout_count"] = 0;
        return true;
    } else {
        return false;
    }
}

SyncBasedDiscovery.prototype.notifyObserver = function 
  (name, msgType, msg)
{
    this.observer.onStateChanged(name, msgType, msg);
    return;
}

SyncBasedDiscovery.prototype.onRegisterFailed = function
  (prefix)
{
    console.log("Prefix registration failed: " + prefix.toUri());
    return;
}

SyncBasedDiscovery.prototype.onEntityDataNotFound = function 
  (prefix, interest, face, interestFilterId, filter)
{
    var name = interest.getName().toUri();
    if (name in this.hostedObjects) {
        var content = this.serializer.serialize(this.hostedObjects[name]);
        var data = new Data(new Name(name));

        data.setContent(content);
        // Interest issuer should not ask for mustBeFresh in this case, for now
        data.getMetaInfo().setFreshnessPeriod(this.entityDataFreshnessPeriod);

        var self = this;
        this.keyChain.sign(data, this.certificateName, function() {
            self.memoryContentCache.add(data);
        });
    }
    
    return;
}

File: framework/ndn_iot_python/ndn_iot_python/discovery/sync_based_discovery.py
import time
import sys
import logging
import random
import hashlib

from pyndn import Name, Data, Interest, Face
from pyndn.security import KeyChain
from pyndn.util import MemoryContentCache

try:
    import asyncio
except ImportError:
    import trollius as asyncio

# Generic sync-based discovery implementation
class SyncBasedDiscovery(object):
    """
    Sync (digest exchange) based name discovery.
    Discovery maintains a list of discovered, and a list of hosted objects.
    Calls observer.onStateChanged(name, msgType, msg) when an entity is discovered or removed.
    Uses serializer.serialize(entityObject) to serialize a hosted entity's entityInfo into string.

    :param face:
    :type face: Face
    :param keyChain: 
    :type keyChain: KeyChain
    :param certificateName:
    :type certificateName: Name
    :param syncPrefix:
    :type syncPrefix: Name
    :param observer:
    :type observer: ExternalObserver
    :param serializer:
    :type serializer: EntitySerializer
    """
    def __init__(self, face, keyChain, certificateName, syncPrefix, observer, serializer, 
      syncDataFreshnessPeriod = 4000, initialDigest = "00", syncInterestLifetime = 4000, syncInterestMinInterval = 500,
      timeoutCntThreshold = 3, maxResponseWaitPeriod = 2000, minResponseWaitPeriod = 400, entityDataFreshnessPeriod = 10000):
        self._face = face
        self._keyChain = keyChain
        self._syncPrefix = Name(syncPrefix)

        self._objects = dict()
        self._hostedObjects = dict()

        self._memoryContentCache = MemoryContentCache(self._face)
        self._certificateName = Name(certificateName)

        self._currentDigest = initialDigest
        self._syncDataFreshnessPeriod = syncDataFreshnessPeriod
        self._initialDigest = initialDigest
        self._syncInterestLifetime = syncInterestLifetime

        self._syncInterestMinInterval = syncInterestMinInterval
        self._timeoutCntThreshold = timeoutCntThreshold
        self._entityDataFreshnessPeriod = entityDataFreshnessPeriod
        # TODO: policy manager etc setup

        #self._maxResponseWaitPeriod = maxResponseWaitPeriod
        #self._minResponseWaitPeriod = minResponseWaitPeriod

        self._observer = observer
        self._serializer = serializer
        self._numOutstandingInterest = 0

        return

    """
    Public facing interface
    """
    def start(self):
        """
        Starts the discovery
        """
        interest = Interest(Name(self._syncPrefix).append(self._initialDigest))
        interest.setMustBeFresh(True)
        interest.setInterestLifetimeMilliseconds(self._syncInterestLifetime)
        self._face.expressInterest(interest, self.onSyncData, self.onSyncTimeout)
        self._numOutstandingInterest += 1
        if __debug__:
            print("Express interest: " + interest.getName().toUri())
        return

    def stop(self):
        """
        Stops the discovery
        """
        # TODO: interest expression and data response flag
        self._memoryContentCache.unregisterAll()
        return

    def getHostedObjects(self):
        return self._hostedObjects

    def getObjects(self):
        return self._objects

    def publishObject(self, name, entityInfo):
        """
        Adds another object and registers prefix for that object's name

        :param name: the object's name string
        :type name: str
        :param entityInfo: the application given entity info to describe this object name with
        :type entityInfo: EntityInfo
        """
        # If this is the first object we host, we register for sync namespace: meaning a participant not hosting anything 
        #   is only "listening" for sync, and will not help in the sync process
        if len(self._hostedObjects.keys()) == 0:
            self._memoryContentCache.registerPrefix(self._syncPrefix, self.onRegisterFailed, self.onSyncInterest)
        if self.addObject(name, False):
            # Do not add itself to contentCache if its currentDigest is "00".
            if self._currentDigest != self._initialDigest:
                self.contentCacheAddSyncData(Name(self._syncPrefix).append(self._currentDigest))
            
            self.updateDigest()
            
            interest = Interest(Name(self._syncPrefix).append(self._currentDigest))
            interest.setInterestLifetimeMilliseconds(self._syncInterestLifetime)
            interest.setMustBeFresh(True)
            
            self._face.expressInterest(interest, self.onSyncData, self.onSyncTimeout)
            self._numOutstandingInterest += 1

            self._hostedObjects[name] = entityInfo
            self.contentCacheAddEntityData(name, entityInfo)
            self.notifyObserver(name, "ADD", "")
            # TODO: should the user configure this prefix as well?
            self._memoryContentCache.registerPrefix(Name(name), self.onRegisterFailed, self.onEntityDataNotFound)
        else:
            if __debug__:
                print("Item with this name already added")
        return

    def removeHostedObject(self, name):
        """
        Removes a locally hosted object

        :param name: the object's name string
        :type name: str
        :return: whether removal's successful or not
        :rtype: bool
        """
        if name in self._hostedObjects:
            del self._hostedObjects[name]
            if len(self._hostedObjects) == 0:
                self._memoryContentCache.unregisterAll()
            if self.removeObject(name):    
                return True
            else:
                if __debug__:
                    print("Hosted item not in objects list")
                return False
        else:
            return False

    """
    Internal functions
    """
    def contentCacheAddEntityData(self, name, entityInfo):
        content = self._serializer.serialize(entityInfo)
        data = Data(Name(name))

        data.setContent(content)

        data.getMetaInfo().setFreshnessPeriod(self._entityDataFreshnessPeriod)
        self._keyChain.sign(data, self._certificateName)
        self._memoryContentCache.add(data)
        print "added entity to cache: " + data.getName().toUri() + "; " + data.getContent().toRawStr()

    def contentCacheAddSyncData(self, dataName):
        sortedKeys = sorted(self._objects.keys())
        content = ""
        for key in sortedKeys:
            content += key + "\n"
        content.strip()

        data = Data(Name(dataName))
        
        data.setContent(content)
        data.getMetaInfo().setFreshnessPeriod(self._syncDataFreshnessPeriod)
        self._keyChain.sign(data, self._certificateName)
        # adding this data to memoryContentCache should satisfy the pending interest
        self._memoryContentCache.add(data)

    def onSyncInterest(self, prefix, interest, face, interestFilterId, filter):
        if interest.getName().size() != self._syncPrefix.size() + 1:
            # Not an interest for us
            return
        digest = interest.getName().get(-1).toEscapedString()
        self.updateDigest()
        if digest != self._currentDigest:
            # Wait a random period before replying; rationale being that "we are always doing ChronoSync recovery...this is the recovery timer but randomized"
            # Consider this statement: we are always doing ChronoSync recovery
            # TODO: this has the problem of potentially answering with wrong data, there will be more interest exchanges needed for the lifetime duration of one wrong answer
            # Consider appending "answerer" as the last component of data name?
            # TODO2: don't see why we should wait here

            self.replySyncInterest(interest, digest)
            #dummyInterest = Interest(Name("/local/timeout1"))
            #dummyInterest.setInterestLifetimeMilliseconds(random.randint(self._minResponseWaitPeriod, self._maxResponseWaitPeriod))
            #self._face.expressInterest(dummyInterest, self.onDummyData, lambda a : self.replySyncInterest(a, digest))
        return

    def replySyncInterest(self, interest, receivedDigest):
        self.updateDigest()
        if receivedDigest != self._currentDigest:
            # TODO: one participant may be answering with wrong info: scenario: 1 has {a}, 2 has {b}
            # 2 gets 1's {a} and asks again before 1 gets 2's {b}, 2 asks 1 with the digest of {a, b}, 1 will 
            # create a data with the content {a} for the digest of {a, b}, and this data will be able to answer
            # later steady state interests from 2 until it expires (and by which time 1 should be updated with
            # {a, b} as well)
            self.contentCacheAddSyncData(Name(self._syncPrefix).append(receivedDigest))
        return

    def onSyncData(self, interest, data):
        # TODO: do verification first
        if __debug__:
            print("Got sync data; name: " + data.getName().toUri() + "; content: " + data.getContent().toRawStr())
        content = data.getContent().toRawStr().split('\n')
        self._numOutstandingInterest -= 1
        for itemName in content:
            if itemName not in self._objects:
                if itemName != "":
                    self.onReceivedSyncData(itemName)

        # Hack for re-expressing sync interest after a short interval
        dummyInterest = Interest(Name("/local/timeout"))
        dummyInterest.setInterestLifetimeMilliseconds(self._syncInterestLifetime)
        self._face.expressInterest(dummyInterest, self.onDummyData, self.onSyncTimeout)
        self._numOutstandingInterest += 1
        return

    def onSyncTimeout(self, interest):
        newInterest = Interest(Name(self._syncPrefix).append(self._currentDigest))
        newInterest.setInterestLifetimeMilliseconds(self._syncInterestLifetime)
        newInterest.setMustBeFresh(True)
        self._numOutstandingInterest -= 1
        print "re-expressing: " + newInterest.getName().toUri()

        if self._numOutstandingInterest <= 0:
            self._face.expressInterest(newInterest, self.onSyncData, self.onSyncTimeout)
            self._numOutstandingInterest += 1
        return

    """
    Handling received sync data: express entity interest
    """
    def onReceivedSyncData(self, itemName):
        interest = Interest(Name(itemName))
        interest.setInterestLifetimeMilliseconds(4000)
        interest.setMustBeFresh(False)
        self._face.expressInterest(interest, self.onEntityData, self.onEntityTimeout) 
        return

    def onEntityTimeout(self, interest):
        print "Item interest times out: " + interest.getName().toUri()
        return

    def onEntityData(self, interest, data):
        self.addObject(interest.getName().toUri(), True)
        self.notifyObserver(interest.getName().toUri(), "ADD", "");

        dummyInterest = Interest(Name("/local/timeout"))
        dummyInterest.setInterestLifetimeMilliseconds(4000)
        self._face.expressInterest(dummyInterest, self.onDummyData, lambda a : self.expressHeartbeatInterest(a, interest))
        return

    def expressHeartbeatInterest(self, dummyInterest, entityInterest):
        newInterest = Interest(entityInterest)
        newInterest.refreshNonce()
        self._face.expressInterest(newInterest, self.onHeartbeatData, self.onHeartbeatTimeout)

    def onHeartbeatData(self, interest, data):
        self.resetTimeoutCnt(interest.getName().toUri())
        dummyInterest = Interest(Name("/local/timeout"))
        dummyInterest.setInterestLifetimeMilliseconds(4000)
        self._face.expressInterest(dummyInterest, self.onDummyData, lambda a : self.expressHeartbeatInterest(a, interest))

    def onHeartbeatTimeout(self, interest):
        if self.incrementTimeoutCnt(interest.getName().toUri()):
            print "Remove: " + interest.getName().toUri() + " because of consecutive timeout cnt exceeded"
        else:
            newInterest = Interest(interest.getName())
            newInterest.setInterestLifetimeMilliseconds(4000)
            self._face.expressInterest(newInterest, self.onHeartbeatData, self.onHeartbeatTimeout)

    def onDummyData(self, interest, data):
        if __debug__:
            print("Unexpected reply to dummy interest: " + data.getContent().toRawStr())
        return

    def expressSyncInterest(self, interest):
        self.onSyncTimeout(interest)
        return

    def addObject(self, name, update):
        if name in self._objects:
            return False
        else:
            self._objects[name] = {"timeout_count": 0}
            if update:
                self.updateDigest()
            return True

    def removeObject(self, name):
        if name in self._objects:
            del self._objects[name]
            self.notifyObserver(name, "REMOVE", "")
            self.contentCacheAddSyncData(Name(self._syncPrefix).append(self._currentDigest))
            self.updateDigest()
            return True
        else:
            return False

    def updateDigest(self):
        # TODO: for now, may change the format of the list encoding for easier cross language compatibility
        if len(self._objects) > 0:
            m = hashlib.sha256()
            for item in sorted(self._objects.keys()):
                m.update(item)
            self._currentDigest = str(m.hexdigest())
        else:
            self._currentDigest = self._initialDigest
        return

    def incrementTimeoutCnt(self, name):
        if name in self._objects:
            self._objects[name]["timeout_count"] += 1
            if self._objects[name]["timeout_count"] >= self._timeoutCntThreshold:
                return self.removeObject(name)
            else:
                return False
        else:
            return False

    def resetTimeoutCnt(self, name):
        if name in self._objects:
            self._objects[name]["timeout_count"] = 0
            return True
        else:
            return False

    def notifyObserver(self, name, msgType, msg):
        self._observer.onStateChanged(name, msgType, msg)
        return

    def onRegisterFailed(self, prefix):
        if __debug__:
            print("Prefix registration failed: " + prefix.toUri())
        return

    def onEntityDataNotFound(self, prefix, interest, face, interestFilterId, filter):
        name = interest.getName().toUri()
        if name in self._hostedObjects:
            content = self._serializer.serialize(self._hostedObjects[name])
            data = Data(Name(name))
            data.setContent(content)
            data.getMetaInfo().setFreshnessPeriod(self._entityDataFreshnessPeriod)
            self._keyChain.sign(data, self._certificateName)
            self._face.putData(data)
        return
